### DISCUSSION TITLE
How to build on windows?

### DISCUSSION THREAD

**Discussion by scarletquasar** [ggml-org/whisper.cpp#85] [Q&A]
-

**Comment by kevin01881** [comment]
It works perfectly through WSL on Windows 10. Install Ubuntu ('wsl --install' command in Powershell). Then install the make and build-essential packages in your Ubuntu instance, and you're all set.

From there, you can follow the steps written by @ggerganov in the readme, as if you were on Linux (well, you actually *are* using a Linux instance at that point, albeit a virtual one 🙂)

**Comment by ggerganov** [comment]
Another option is to get the latest executable from the Actions page:

https://github.com/ggerganov/whisper.cpp/actions

**Reply by vitacon** [reply]
Well, I managed to compile "main.exe" in Visual Studio 2022 running cmake:
`"c:\Program Files\Microsoft Visual Studio\2022\Preview\Common7\IDE\CommonExtensions\Microsoft\CMake\CMake\bin\cmake.exe" -S . -B ./build`

It created a bunch of different project files and I could open .\build\ALL_BUILD.vcxproj and compile it from VS IDE.

However, I would still like to get the "proper" latest automated build, but all I can see on Actions page are logs from the compilation process:

```
Run actions/upload-artifact@v1
Uploading artifact 'whisper-bin-x64' from 'D:\a\whisper.cpp\whisper.cpp\build\bin\RelWithDebInfo' for run #300
Uploading 12 files
File upload complete.
Uploaded '61310873' bytes from 'D:\a\whisper.cpp\whisper.cpp\build\bin\RelWithDebInfo' to server
Associated artifact whisper-bin-x64 (1821441) with run #300
```

I suppose I am not looking in the right place? It would be also easier if the exact link was mentioned on the main page (readme.md) but I can't find it there either... =(

**Reply by jake1271** [reply]
Do you know where on the Actions page the Windows executable can be downloaded? I went to the last page, currently 38, and don't see any execs to download.  Sorry for being so dense.

**Reply by rovashan** [reply]
@jake1271 Click on one of the successful runs then you will see the artifacts at the bottom.

**Reply by jake1271** [reply]
Ok thanks, so any successful run is ok? I took this one from tag master: https://github.com/ggerganov/whisper.cpp/actions/runs/4567265807

**Reply by rovashan** [reply]
Yes, that should be fine. Typically the newer the better.

**Reply by xsploit** [reply]
> Another option is to get the latest executable from the Actions page:
> 
> https://github.com/ggerganov/whisper.cpp/actions

can you still do this?

**Comment by RYucel** [comment]
At the end of actions page, you will find executables for Windows. To
download, you need to login your GitHub account.

**Reply by vitacon** [reply]
Thanks! I finally found it. =}
My "blindness" was caused by the panel with "build.yml". I use mouse wheel for scrolling so when the mouse cursor reached the panel it stopped scrolling and I never got to see the rest of the page and I even did not notice I was missing something.

**Reply by rovashan** [reply]
Can you provide the command to run the Windows executable please?

**Reply by vitacon** [reply]
Run the process to convert .wav to .srt?
The simplest way is this:
`main -f input.wav`

But the audio must be of 16 kHz and you must place the right model file to "models" subfolder first.

**Reply by rovashan** [reply]
I would like it to transcribe using the mic input rather than from a file.

I tried running the below command but it's printing nonsense.
`stream -m "C:\repo\whisper.cpp\models\ggml-base.en.bin" -t 8 --step 500 --length 5000`

There is also this warning message that comes up so maybe it needs a faster machine.
`SDL_main: WARNING: cannot process audio fast enough, dropping audio ...`

**Reply by vitacon** [reply]
Sorry, I have no experience with it. But if your machine is not fast enough you could try "tiny" model. Smaller models are faster.

**Reply by rovashan** [reply]
It's better with the tiny model but still not great so I will try a different machine.

**Comment by Const-me** [comment]
Technically, File / Open / CMake works OK.

Practically, the .exe made that way doesn’t use AVX2 or even AVX, and takes 7.5 minutes to process `samples/jfk.wav` using `medium.en` model.

**Reply by RndyP** [reply]
#define __AVX__
The difference in processing time is HUGE.

**Reply by ggerganov** [reply]
@Const-me Do you mean that the build with CMake does not enable the AVX flags? (I don't have Windows to test)
If that is the case, then it is a bug. The following should enable AVX on Windows:

https://github.com/ggerganov/whisper.cpp/blob/832b4f34c9a155352c9f4c121faf92dcf759f86d/CMakeLists.txt#L148-L152

**Reply by vitacon** [reply]
I can confirm the current CMake behaviour.

Your build works with AVX:
`system_info: n_threads = 11 / 12 | AVX = 1 | AVX2 = 1 | AVX512 = 0 | NEON = 0 | F16C = 0 | FP16_VA = 0 | WASM_SIMD = 0 | BLAS = 1 |`

My today's local CMake build does not:
`system_info: n_threads = 11 / 12 | AVX = 0 | AVX2 = 0 | AVX512 = 0 | NEON = 0 | F16C = 0 | FP16_VA = 0 | WASM_SIMD = 0 | BLAS = 0 |`

However my yesterday's build made from IDE does this:
`system_info: n_threads = 11 / 12 | AVX = 1 | AVX2 = 1 | AVX512 = 0 | NEON = 0 | F16C = 0 | FP16_VA = 0 | WASM_SIMD = 0 | BLAS = 0 |`

Does BLAS affect performance?

**Reply by ggerganov** [reply]
Hmm, I somehow missed that we are building the Windows packages with BLAS enabled.
My experience on the couple of Linux machines that I have is that without BLAS it is faster. So I would recommend disabling it and see if the performance improves (i.e. the build that you made from your IDE).

If you want to enable BLAS in your IDE, you have to run CMake with `-DWHISPER_SUPPORT_OPENBLAS=ON` and have OpenBLAS installed properly.

Btw, the Github action uses CMake to produce the binaries and they have AVX enabled (based on your comment) so it does work correctly. Try following the steps from the .yml file:

https://github.com/ggerganov/whisper.cpp/actions/runs/3650728538/workflow#L117-L185

**Reply by vitacon** [reply]
I've got an additional and probably rather tricky question:

I was asked to run Whisper.cpp on an older computer and it does not work. I though it is because of the special instructions and other stuff so I tried to run it on several other computers to get a better picture. i5-4xxx and higher works fine, i5-860 and similar PCs crash with an error message, but the "original" PC uses i7-3770 and it simply shuts down without any error message which is rather strange.

It shuts down BEFORE it should display system_info so I thought it could be solved by disabling AVX2 (i7-3770 should support at least AVX). Unfortunately even the CMake build that does not use any special stuff  (i.e. it displays AVX2 = 0 on my Ryzen) shuts down at the same moment.

Do you have any explanation for this? And is there any way around it? (Skip some HW/CPU check and force just AVX support?)

**Reply by vitacon** [reply]
> My experience on the couple of Linux machines that I have is that without BLAS it is faster.

I have not measured the differences yet. =} I suppose I could stick to "my build" if the difference are noticeable.

**Reply by vitacon** [reply]
After some quick testing it is apparent that the differences ARE noticeable. =)

```
CMake build         | 486330.88 ms
whisper-bin-x64 (1) |  88582.91 ms
whisper-bin-x64 (2) |  96475.12 ms
my IDE build (1)    |  22898.04 ms
my IDE build (2)    |  22691.02 ms
```

Actually, I expected the app to use only the fastest option that is available. I.e. I though it would run system_info and then it would enable only the best available option.

**Reply by RndyP** [reply]
On Windows I used the IDE to simply create a new C++ command line project and then downloaded the Whisper source files into the new Whisper folder, then added the files to the project. It was pretty easy to set up. Like I said above, I had to add a #define __AVX__ to ggml.h, although this could have been set in the IDE instead. Performance improvement was huge; I was quite surprised. Realtime transcribing is impossible without it.

![Screenshot 2022-12-08 180511](https://user-images.githubusercontent.com/78916608/206592593-775f8f2e-111c-4187-ba66-abc248a96ac9.jpg)

**Reply by RndyP** [reply]
Here's an example on my Dell laptop with a i7-8550U CPU:

![Screenshot 2022-12-08 182246](https://user-images.githubusercontent.com/78916608/206594526-c453e0e8-0cb7-471b-be3f-256a1b229c02.jpg)

**Reply by Const-me** [reply]
> @Const-me Do you mean that the build with CMake does not enable the AVX flags?

Indeed. Not even AVX1 was enabled in the cmake builds.

> The following should enable AVX on Windows:
> https://github.com/ggerganov/whisper.cpp/blob/832b4f34c9a155352c9f4c121faf92dcf759f86d/CMakeLists.txt#L148-L152

That didn’t work for some reason. At least not in Visual Studio 2022.
Note that for optimal performance, the DLL project also needs `/GL` compiler switch, and `/LTCG` linker switch.

See the complete projects in my fork: https://github.com/Const-me/whisper.cpp/tree/master/Windows

BTW, consider merging that fork, these [two](https://github.com/ggerganov/whisper.cpp/commit/5e4ef66a2ee43c89cafd1532df14ccb90807a29b) [commits](https://github.com/ggerganov/whisper.cpp/commit/5a6bc23739efc5c3be1b5f2adfece0b26fbdbbb7) improved performance by about 20% on my computer.

**Reply by xuanyinmu** [reply]
> Here's an example on my Dell laptop with a i7-8550U CPU:
> 
> ![Screenshot 2022-12-08 182246](https://user-images.githubusercontent.com/78916608/206594526-c453e0e8-0cb7-471b-be3f-256a1b229c02.jpg)

Hello, i'm the beginner.
I want to create a new project in VS to run whispercpp like you. I haven’t changed anything,just copy the main.cpp code to ASRTCPP.cpp, but I always report an error that the vector subscript is out of bounds. I don’t know what’s going on. If you can help solve it, thank you very much.
![运行截图](https://user-images.githubusercontent.com/60185303/207008446-f805af25-e673-4269-8966-bf377ff7ac17.jpg)

![main参数](https://user-images.githubusercontent.com/60185303/207008469-9142d96e-f1cb-4bc6-a583-c002cda2fef9.jpg)

![代码结构](https://user-images.githubusercontent.com/60185303/207009800-d1496496-da61-4105-85be-82cdff715125.jpg)

![代码文件](https://user-images.githubusercontent.com/60185303/207009830-1c6985c4-9a22-4dbe-a328-edc512fa1b14.jpg)

**Reply by vitacon** [reply]
> Hmm, I somehow missed that we are building the Windows packages with BLAS enabled. My experience on the couple of Linux machines that I have is that without BLAS it is faster. So I would recommend disabling it and see if the performance improves (i.e. the build that you made from your IDE).

I am not sure if you tried to do something about it but your Windows GitHub build still reports that BLAS is enabled so the build is rather slow... 😿

**Reply by ggerganov** [reply]
There is now also a non-Blas Windows build.
But don't expect major performance improvements - probably few tens of per cent.

**Reply by vitacon** [reply]
Thanks! It will save me some time needed for manual compilation.

The difference is really noticeable - at least on my CPU (Ryzen 5 3600).
These are numbers with a random vocal sample and Medium model:

```
"my build"     218 596 ms
BLAS disabled  218 038 ms
BLAS enabled   510 956 ms

```

With Tiny model the ratio was even bigger (about 1:4).

**Comment by siqueira111** [comment]
How can I compile it with cl.exe? what is the lib that I need to link it to?

**Reply by siqueira111** [reply]
Nvm, it was a silly question... I was supposed to copy whisper.lib inside the source code path...
and then use 
#pragma comment(lib, "whisper")

**Comment by Patch4747** [comment]
Can I use the whisper.dll asset in a C# project? I'm not sure how to know what the exported methods are.

**Reply by AmSmart** [reply]
Sure, why not! You can also checkout the [Whisper.net project](https://github.com/sandrohanea/whisper.net)

**Comment by Kreijstal** [comment]
You can read the PKGBUILDs from the msys2 project

**Comment by ulatekh** [comment]
Building with CMake and Visual Studio 2019 worked fine for me.
